### ISSUE TITLE
High RAM usage on idle with prometheus 7.0

### ISSUE THREAD

**Issue by Jakimar** [prometheus-net/prometheus-net#394]
After upgrading to prometheus-net.AspNetCore "7.0.0" i'm noticing high usage of RAM on my web api projects even on idle but only if you open metrics page at least once. Same behavior with Sample.Web project in this repo. At start RAM usage is not high, but it's continues to grow with time, so with "7.0.0" version it's begins to slowly grow after launch from 20 mb up to 70-100mb after 30-60min in my case. Problem disappear if you downgrade to "6.0.0"

Steps to reproduce:
1. Create project with code below.
2. Open metrics page "*/metrics" couple of times.

Code:
```
using Prometheus;

var builder = WebApplication.CreateBuilder(args);

var app = builder.Build();

app.UseMetricServer();
app.UseHttpMetrics();

app.MapGet("/", () => "Hello World!");

app.Run();

```

csproj:
```
<Project Sdk="Microsoft.NET.Sdk.Web">

    <PropertyGroup>
        <TargetFramework>net6.0</TargetFramework>
        <Nullable>enable</Nullable>
        <ImplicitUsings>enable</ImplicitUsings>
    </PropertyGroup>

    <ItemGroup>
        <PackageReference Include="prometheus-net.AspNetCore" Version="7.0.0" />
    </ItemGroup>

</Project>
```

**Update**
So i finded advice from @sandersaares in issue #392 to disable event counter integration and it's also does the trick and abnormal RAM usage stops. 

workaround: 
```
Metrics.SuppressDefaultMetrics(new SuppressDefaultMetricOptions
{
    SuppressEventCounters = true
});
```

**Comment by BigDaddy1337** [comment]
Same here, this is my app memory usage before/after upgrade to prometheus 7.0.0 with enabled EventCounters integration. 

<img width="1596" alt="image" src="https://user-images.githubusercontent.com/10048426/212854312-07bf5c21-3066-47c4-8216-d893eadec474.png">

**Comment by matteocontrini** [comment]
I'm experiencing this memory leak too. Could you please clarify what metrics do I lose if I disable event counters, exactly? Thanks!

**Comment by sandersaares** [comment]
The affected metrics are those that come from registered .NET EventCounters.

In most apps, you will see `system_runtime_*`, `system_net_*` and `microsoft_aspnetcore_*` metrics as the affected default metrics. Depending on the installed set of modules, there may also be others, as v7.0.0 by default exposes all event counters. There will be a small change in this for v8.0.0 to only expose the most generic set of 3 (what I listed here) and require opt-in for others (to ensure optimal default configuration performance in other ways not related to this issue).

**Comment by sandersaares** [comment]
Investigation by .NET team revealed that this memory usage is completely normal and is, in fact, just garbage that has not yet been collected by the garbage collector. This might be because the app is very lightly loaded and experiences no memory pressure - in which case garbage collection may be quite rare.

You can verify this by adding the following (non-production-grade) code to your app, to force it to constantly collect garbage. You should see memory usage go flat with this.

```csharp
Thread t = new Thread(GCCollectThread)
{
    IsBackground = true
};
t.Start();

void GCCollectThread(object? obj)
{
    while (true)
    {
        Thread.Sleep(2000);
        GC.Collect();
        GC.WaitForPendingFinalizers();
        GC.Collect();
        Console.Write(".");
    }
}
```

Based on this, it looks like there is no problem - the system is behaving as intended.

@Jakimar, @BigDaddy1337, @matteocontrini Can you confirm that you see no rise in memory usage with forced garbage collection?

**Comment by sandersaares** [comment]
As the issue looks scary (even though it seems to be harmless), in the latest 8.0.0 preview build published today, I have added an experimental mitigation - would be interesting to hear if behavior looks different for anyone with the latest prerelease package.

**Comment by Jakimar** [comment]
Yeah, with forced garbage collection memory rise is no longer visible. Also i saw slight increase of CPU usage with 7.0.0, so here are some graphs for idle from one of my application in test enviroment after getting metrics couple of time:

**Version "6.0.0"**:
CPU:
![Ibs-6-Cpu-15m](https://user-images.githubusercontent.com/116259451/215330199-758f54cf-0d0e-4a5f-a3d8-0d2d6956ded9.PNG)
Memory:
![Ibs-6-Memory15m](https://user-images.githubusercontent.com/116259451/215330422-8184895a-5152-4284-99ae-79d47f5dd800.PNG)


**Version "7.0.0"**:
CPU:
![Ibs-7-Cpu-15m](https://user-images.githubusercontent.com/116259451/215330500-8279dff1-4c2c-46dd-bdc6-b8a7d534b756.PNG)
Memory:
![Ibs-7-Memory15m](https://user-images.githubusercontent.com/116259451/215330708-54b1baec-902d-49b1-8bf5-7e4b20c09a5a.PNG)
For 7.0.0 i waited **24+ hours** before collecting data because with event counter integration at start memory spikes around 1.5x times. 

**Version "8.0.0-pre-230127154206-9ec9e9b"**:
CPU :
![Ibs-8-Cpu-15m](https://user-images.githubusercontent.com/116259451/215330841-5241013b-7bdb-4192-b4f2-8479913b227f.PNG)
Memory:
![Ibs-8-Memory15m](https://user-images.githubusercontent.com/116259451/215330851-d7c929ef-dc9f-40a0-a4a4-6e6ae491a3a6.PNG)


As you can see with 7.0.0 memory is always higher, for idle maybe it's not so much so something around 10-20%(assuming launch spike is already GC'ed), but in production with small preasure around 10 RPS i saw memory increase from 150mb avg to 250mb avg. I will try 8.0.0 preview in production next week and will let it run for couple of days to see memory difference in long run.

For CPU version 7.0.0 was harsh, for example i have around 60 instanses of different apps in test and if i upgrade all them to 7.0.0 based on graphs above i can expect addtional constant usage of 3-5% CPU on my 8-CPU server setup. With current 8.0.0 preview it's nice to see it no longer a problem and CPU is almost as low as 6.0.0.

**Comment by Jakimar** [comment]
After a week of observations i can't see any problem with version "8.0.0-pre-230127154206-9ec9e9b" both memory and CPU looks totally fine.

**Comment by matteocontrini** [comment]
@sandersaares regarding this issue we've talked about a few months ago, I'm still suppressing event counters in my code (even after upgrading to 8.0) but I was wondering what the actual effect of that is.

I've tried removing the suppress thing but I don't see anything changing in the list of exported metrics. The only new metric I see is `prometheus_net_eventcounteradapter_sources_connected_total` with a value of 4 or 6 depending on the application, but no actual new metric.

Am I missing something? I was under the impression that by default this library collected some metrics from EventCounter, which is what was causing this issue. I'm using the latest version of this library on ASP.NET Core 8.0. Any explanation? Thanks.

**Comment by jbojcic1** [comment]
> I've tried removing the suppress thing but I don't see anything changing in the list of exported metrics. The only new metric I see is `prometheus_net_eventcounteradapter_sources_connected_total` with a value of 4 or 6 depending on the application, but no actual new metric.
> 
> Am I missing something? I was under the impression that by default this library collected some metrics from EventCounter, which is what was causing this issue. I'm using the latest version of this library on ASP.NET Core 8.0. Any explanation? Thanks.

I noticed the same thing on version 8.2. I actually want event counter metrics and according to the docs they should be added by default (all well known counters as doc say) but I don't see any of them. I also just see `prometheus_net_eventcounteradapter_sources_connected_total`
